3/assets/3.6-finished-code.rs

```diff
@@ -0,0 +1,315 @@
+                gen: cmp::max(kitty_1.gen, kitty_2.gen) + 1,
+            };
+
+            Self::_mint(sender, random_hash, new_kitty)?;
+
+            <Nonce<T>>::mutate(|n| *n += 1);
+
+            Ok(())
+        }
+
+        fn create_auction(origin, kitty_id: T::Hash, min_bid: T::Balance, expiry_in_minutes: u64) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner == sender, "You can't set an auction for a cat you don't own");
+
+            let expiry = <system::Module<T>>::block_number() + T::BlockNumber::sa(expiry_in_minutes*12);
+            ensure!(expiry <= <system::Module<T>>::block_number() + Self::auction_duration_limit(), "The auction duration should be less than 5 days");
+
+            let new_auction = Auction {
+                kitty_id,
+                kitty_owner: owner,
+                expiry,
+                min_bid,
+                high_bid: min_bid,
+                high_bidder: sender,
+            };
+
+            <KittyAuction<T>>::insert(kitty_id, &new_auction);
+            <Auctions<T>>::insert(expiry, vec![new_auction.clone()]);
+
+            Self::deposit_event(RawEvent::Auction(kitty_id, min_bid, expiry));
+
+            Ok (())
+        }
+
+        fn bid_auction(origin, kitty_id: T::Hash, bid: T::Balance) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner != sender, "You can't bid for your own cat");
+
+            let mut auction = Self::auction_of(kitty_id).ok_or("No auction for this cat")?;
+            ensure!(<system::Module<T>>::block_number() < auction.expiry, "This auction is expired.");
+
+            ensure!(bid > auction.min_bid, "Your bid has to be greater than the minimum bid.");
+
+            ensure!(bid > auction.high_bid, "Your bid has to be greater than the highest bid.");
+
+            auction.high_bid = bid;
+            auction.high_bidder = sender.clone();
+            <KittyAuction<T>>::insert(kitty_id, &auction);
+            <Auctions<T>>::insert(auction.expiry, vec![auction.clone()]);
+
+            Self::deposit_event(RawEvent::Bid(kitty_id, auction.high_bid, auction.high_bidder));
+
+            Ok (())
+        }
+
+        fn on_finalise() {
+            let auctions = Self::auctions_expire_at(<system::Module<T>>::block_number());
+            for auction in &auctions {
+                if <system::Module<T>>::block_number() >= auction.expiry && auction.kitty_owner != auction.high_bidder {
+                    let _ = Self::_transfer_from(auction.kitty_owner.clone(), auction.high_bidder.clone(), auction.kitty_id);
```

Comment:
You need to make checks ahead of time to ensure neither will fail. Both have some amount of `verify first, write last`, and if one succeeds, and the other fails, you will be hurting someone on your chain. (someone wont get the kitty they paid for, or someone wont get their money)

This will be less of an issue when you change to use the escrow method, since we can ensure that the user will get their money, so we just need to make sure the kitty will transfer correctly